pymc3/step_methods/smc_ABC.py

```diff
@@ -0,0 +1,418 @@
+    """
+    Calculate trace covariance matrix based on importance weights.
+    """
+    cov = np.cov(
+        np.squeeze(posterior_array),
+        aweights=weights.ravel(),
+        bias=False,
+        rowvar=0)
+    if np.isnan(cov).any() or np.isinf(cov).any():
+        raise ValueError(
+            'Sample covariances not valid! Likely "chains" is too small!')
+    return np.atleast_2d(cov)
+
+
+def _tune(acc_rate):
+    """
+    Tune adaptively based on the acceptance rate.
+
+    Parameters
+    ----------
+    acc_rate: float
+        Acceptance rate of the Metropolis sampling
+
+    Returns
+    -------
+    scaling: float
+    """
+    # a and b after Muto & Beck 2008 .
+    a = 1. / 9
+    b = 8. / 9
+    return (a + b * acc_rate) ** 2
+
+
+def _posterior_to_trace(posterior, model):
+    """
+    Save results into a PyMC3 trace
+    """
+    length_pos = len(posterior)
+    varnames = [v.name for v in model.vars]
+    with model:
+        strace = NDArray(model)
+        strace.setup(length_pos, 0)
+    for i in range(length_pos):
+        strace.record({k: v for k, v in zip(varnames, posterior[i])})
+    return MultiTrace([strace])
+
+
+def get_sum_stats(data, sum_stat=SMC_ABC().sum_stat):
+    """
+    Parameters:
+    -----------
+    data : array
+        Observed or simulated data
+    sum_stat : list
+        List of summary statistics to be computed. Accepted strings are mean, std, var.
+        Python functions can be passed in this argument.
+
+    Returns:
+    --------
+    sum_stat_vector : array
+        Array contaning the summary statistics.
+    """
+    if data.ndim == 1:
+        data = data[:, np.newaxis]
+    sum_stat_vector = np.zeros((len(sum_stat), data.shape[1]))
+
+    for i, stat in enumerate(sum_stat):
+        for j in range(sum_stat_vector.shape[1]):
+            if stat == 'mean':
+                sum_stat_vector[i, j] = data[:, j].mean()
+            elif stat == 'std':
+                sum_stat_vector[i, j] = data[:, j].std()
+            elif stat == 'var':
+                sum_stat_vector[i, j] = data[:, j].var()
+            else:
+                sum_stat_vector[i, j] = stat(data[:, j])
+
+    return np.atleast_1d(np.squeeze(sum_stat_vector))
+
+
+def absolute_difference(a, b):
+    return np.sum(np.abs(a - b))
+
+
+def sum_of_squared_distance(a, b):
+    return np.sum((a - b)**2)
+
+
+def mean_absolute_error(a, b):
+    return np.sum(np.abs(a - b)) / len(a)
+
+
+def mean_squared_error(a, b):
+    return np.sum((a - b)**2) / len(a)
+
+
+def euclidean_distance(a, b):
+    return np.sqrt(np.sum((a - b)**2))
+
+
+def get_distance(func_name):
+    d = {'absolute difference': absolute_difference,
+         'sum of squared_distance': sum_of_squared_distance,
+         'mean absolute_error': mean_absolute_error,
+         'mean squared_error': mean_squared_error,
+         'euclidean': euclidean_distance}
+    for key, value in d.items():
+        if func_name == key:
+            return value
```

Comment:
the last three lines can be simplified to avoid a for loop:
```
if not func_name in d:
    raise KeyError('"{}" was not found in the standard distance metrics. Choose one of {}'.format(func_name, set(d.keys())))
return d[func_name]
```